Jonathan Diaz           24/09/2025

# Tarea - Programación lineal

### Diseño de la Dieta Óptima

Se quiere producir comida para gatos de la manera más barata, no obstante se debe también asegurar que se cumplan los datos requeridos de analisis nutricional. Por lo que se quiere variar la cantidad de cada ingrediente para cumplir con los estandares nutricionales. Los requisitos que se tienen es que en $100$ gramos, se deben tener **por lo menos** $8$ gramos de proteína y $6$ gramos de grasa. Así mismo, no se debe tener más de $2$ gramos de fibra y $0.4$ gramos de sal.  

Los datos nutricionales se pueden obtener de la siguiente tabla:

Ingrediente|Proteína|Grasa|Fibra|Sal
:----|----|----|----|----
Pollo|  10.0%|08.0%|00.1%|00.2%
Carne|  20.0%|10.0%|00.5%|00.5%
Cordero|15.0%|11.0%|00.5%|00.7%
Arroz|  00.0%|01.0%|10.0%|00.2%
Trigo|  04.0%|01.0%|15.0%|00.8%
Gel|    00.0%|00.0%|00.0%|00.0%

Los costos de cada producto son:

Ingrediente|Costo por gramo
:----|----
Pollo|$\$$0.013
Carne|$\$$0.008
Cordero|$\$$0.010
Arroz|$\$$0.002
Trigo|$\$$0.005
Gel|$\$$0.001    

Lo que se busca optimizar en este caso es la cantidad de productos que se debe utilizar en la comida de gato, minimizando el costo total. Para simplificar la notación use las siguientes variables:

+ $x_1:$ Gramos de pollo  
+ $x_2:$ Gramos de carne  
+ $x_3:$ Gramos de cordero  
+ $x_4:$ Gramos de arroz  
+ $x_5:$ Gramos de trigo  
+ $x_6:$ Gramos de gel

La tarea consiste en plantear el problemade programación lineal que permita satisfacer las necesidades alimenticias del gato al tiempo que minimice el costo total y resolverlo con `linprog`.

**Solución:**

Se busca determinar la cantidad de cada ingrediente para una ración de $100\;\mathrm{g}$ de comida para gato, de tal forma que se **minimice el costo total** y se cumplan los requerimientos nutricionales.

**Ecuación principal:**

$$
\min Z = 0.013x_{1}+0.008x_{2}+0.010x_{3}+0.002x_{4}+0.005x_{5}+0.001x_{6}
$$

**Restricciones:**

- **Peso total:**
  $$
  x_{1}+x_{2}+x_{3}+x_{4}+x_{5}+x_{6}=100 \,\text{g}
  $$

- **Proteína:**
  $$
  0.10x_{1}+0.20x_{2}+0.15x_{3}+0.01x_{4}+0.04x_{5}+0.00x_{6}\geq 8\,\text{g}
  $$

- **Grasa:**
  $$
  0.08x_{1}+0.10x_{2}+0.11x_{3}+0.01x_{4}+0.01x_{5}+0.00x_{6}\geq 6\,\text{g}
  $$

- **Fibra:**
  $$
  0.001x_{1}+0.005x_{2}+0.005x_{3}+0.10x_{4}+0.15x_{5}+0.00x_{6}\leq 2\,\text{g}
  $$

- **Sal:**
  $$
  0.002x_{1}+0.005x_{2}+0.007x_{3}+0.002x_{4}+0.008x_{5}+0.000x_{6}\leq 0.4\,\text{g}
  $$

In [ ]:
import numpy as np
import scipy.optimize as opt

c=np.array([.013,.008,.01,.002,.005,.001])
A=np.array([[-.1,-.2,-.15,-0,-.04,-0],
            [-.08,-0.1,-.11,-.01,-.01,-0],
            [.001,.005,.005,.1,.15,0],
            [.002,.005,.007,.002,.008,0]])
b=np.array([-8,-6,2,.4])
Ae=np.array([[1,1,1,1,1,1]])
be=np.array([100])
bounds=(0,None)

opt.linprog(c,A,b,Ae,be,bounds=bounds)

        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: 0.52
              x: [ 0.000e+00  6.000e+01  0.000e+00  0.000e+00  0.000e+00
                   4.000e+01]
            nit: 2
          lower:  residual: [ 0.000e+00  6.000e+01  0.000e+00  0.000e+00
                              0.000e+00  4.000e+01]
                 marginals: [ 6.400e-03  0.000e+00  1.300e-03  3.000e-04
                              3.300e-03  0.000e+00]
          upper:  residual: [       inf        inf        inf        inf
                                    inf        inf]
                 marginals: [ 0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              0.000e+00  0.000e+00]
          eqlin:  residual: [ 0.000e+00]
                 marginals: [ 1.000e-03]
        ineqlin:  residual: [ 4.000e+00  0.000e+00  1.700e+00  1.000e-01]
                 marginals: [-0.000e+00 -7.000e-02 -0.000e+00 -0.000e+00]
 mip

**Resultado:**

El costo mínimo de la dieta es de $0.52\;\mathrm{pesos}$, con una composición de $60\;\mathrm{g}$ de carne y $40\;\mathrm{g}$ de gel.

---

### Problema de distribución de energía eléctrica

La Comisión Federal de Electricidad **(CFE)** dispone de cuatro plantas de generación para satisfacer la demanda diaria eléctrica en cuatro ciudades, Guadalajara, León y Morelia. Las plantas $1$, $2$ y $3$ pueden satisfacer $80$, $40$ y $60$ millones de KW al día respectivamente. Las necesidades de las ciudades de Guadalajara, León y Morelia son de $70$, $40$ y $70$ millones de Kw al día respectivamente.


Los costos asociados al envío de suministro energético por cada millón de Kw entre cada planta y cada ciudad son los registrados en la siguiente tabla.

-|Guadalajara|León|Morelia
:----|----|----|----
Planta 1|5|2|7
Planta 2|3|6|6
Planta 3|6|1|2

Y por último, las restricciones del problema, van a estar dadas por las capacidades de oferta y demanda de cada planta (en millones de KW) y cada ciudad.

Para simplificar la notación use las siguientes variables:

+ $x_1$: Kw (en millones) distribuidos de la Planta 1 a Guadalajara
+ $x_2$: Kw (en millones) distribuidos de la Planta 1 a León
+ $x_3$: Kw (en millones) distribuidos de la Planta 1 a Morelia
+ $x_4$: Kw (en millones) distribuidos de la Planta 2 a Guadalajara
+ $x_5$: Kw (en millones) distribuidos de la Planta 2 a León
+ $x_6$: Kw (en millones) distribuidos de la Planta 2 a Morelia
+ $x_7$: Kw (en millones) distribuidos de la Planta 3 a Guadalajara
+ $x_8$: Kw (en millones) distribuidos de la Planta 3 a León
+ $x_9$: Kw (en millones) distribuidos de la Planta 3 a Morelia

La tarea consiste en plantear el problema de programación lineal que permita satisfacer las necesidades de todas las ciudades al tiempo que minimice los costos asociados a la distribución y resolverlo con `linprog`.

**Solución:**
 La Comisión Federal de Electricidad (CFE) busca **minimizar el costo total de distribución** de energía eléctrica desde tres plantas a tres ciudades, respetando la capacidad de las plantas y la demanda de cada ciudad.

**Ecuación principal:**

$$
\min Z = 5x_{1}+2x_{2}+7x_{3}+3x_{4}+6x_{5}+6x_{6}+6x_{7}+1x_{8}+2x_{9}
$$

**Restricciones:**

- **Capacidad de las plantas:**

  - Planta 1:
    $$
    x_{1}+x_{2}+x_{3}\leq 80\;\mathrm{MW}
    $$

  - Planta 2:
    $$
    x_{4}+x_{5}+x_{6}\leq 40\;\mathrm{MW}
    $$

  - Planta 3:
    $$
    x_{7}+x_{8}+x_{9}\leq 60\;\mathrm{MW}
    $$

- **Demanda de las ciudades:**

  - Guadalajara:
    $$
    x_{1}+x_{4}+x_{7}\geq 70\;\mathrm{MW}
    $$

  - León:
    $$
    x_{2}+x_{5}+x_{8}\geq 40\;\mathrm{MW}
    $$

  - Morelia:
    $$
    x_{3}+x_{6}+x_{9}\geq 70\;\mathrm{MW}
    $$

In [ ]:
import numpy as np
import scipy.optimize as opt

c=np.array([5,2,7,3,6,6,6,1,2])
A=np.array([[1,1,1,0,0,0,0,0,0],
            [0,0,0,1,1,1,0,0,0],
            [0,0,0,0,0,0,1,1,1],
            [-1,0,0,-1,0,0,-1,0,0],
            [0,-1,0,0,-1,0,0,-1,0],
            [0,0,-1,0,0,-1,0,0,-1]])
b=np.array([80,40,60,-70,-40,-70])

opt.linprog(c,A,b)

        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: 540.0
              x: [ 3.000e+01  4.000e+01  1.000e+01  4.000e+01  0.000e+00
                   0.000e+00  0.000e+00  0.000e+00  6.000e+01]
            nit: 6
          lower:  residual: [ 3.000e+01  4.000e+01  1.000e+01  4.000e+01
                              0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              6.000e+01]
                 marginals: [ 0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              6.000e+00  1.000e+00  6.000e+00  4.000e+00
                              0.000e+00]
          upper:  residual: [       inf        inf        inf        inf
                                    inf        inf        inf        inf
                                    inf]
                 marginals: [ 0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              0.000e+00  0.000e+00  0.000e+00  0.0

**Resultado:**

El costo mínimo de distribución es de $540\;\mathrm{millones}$

---

### Fábrica de focos

Una fábrica produce focos incandesentes y los vende a 4.5 u.m., además de focos ahorradores que vende a 6 u.m. cada uno. La producción está limitada a 400 focos incandesentes y 300 focos ahorradores al día, además de no poder producir más de 500 focos en total.

Si la fábrica vende toda su producción, determina cuántos focos incandesentes y cuántos focos ahorradores debe producir para obtener los máximos ingresos posibles y cuáles serían éstos.

**Solución**:

Se busca **maximizar los ingresos** de una fábrica que produce focos incandescentes y ahorradores, con límites en la producción individual y total.

**Ecuación principal:**

$$
\max= 4.5x_{1}+6x_{2}
$$

**Restricciones:**

- **Límite de producción individual:**

  - Incandescentes:
    $$
    x_{1}\leq 400\;\mathrm{focos}
    $$

  - Ahorradores:
    $$
    x_{2}\leq 300\;\mathrm{focos}
    $$

- **Límite de producción total:**
  $$
  x_{1}+x_{2}\leq 500\;\mathrm{focos}
  $$


In [ ]:
import numpy as np
import scipy.optimize as opt

c=np.array([-4.5,-6])
A=np.array([[1,0],
            [0,1],
            [1,1]])
b=np.array([400,300,500])

opt.linprog(c,A,b)

        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: -2700.0
              x: [ 2.000e+02  3.000e+02]
            nit: 1
          lower:  residual: [ 2.000e+02  3.000e+02]
                 marginals: [ 0.000e+00  0.000e+00]
          upper:  residual: [       inf        inf]
                 marginals: [ 0.000e+00  0.000e+00]
          eqlin:  residual: []
                 marginals: []
        ineqlin:  residual: [ 2.000e+02  0.000e+00  0.000e+00]
                 marginals: [-0.000e+00 -1.500e+00 -4.500e+00]
 mip_node_count: 0
 mip_dual_bound: 0.0
        mip_gap: 0.0

**Resultado:**

El ingreso máximo es de $2700\;\mathrm{u.m.}$ produciendo $200\;\mathrm{ incadesentes}$ y $\$300\;\mathrm{ahorradores}$

----

### Problema de compra al mayoreo
Un frutero necesita 16 cajas de naranjas, 5 de plátanos y 20 de manzanas. Dos mayoristas pueden suministrarle para satisfacer sus necesidades, pero sólo venden la fruta en contenedores completos. El mayorista A envía en cada contenedor 8 cajas de naranjas, 1 de plátanos y 2 de manzanas. El mayorista B envía en cada contenedor 2 cajas de naranjas, una de plátanos y 7 de manzanas. Sabiendo que cada contenedor del mayorista A tiene un costo de 150 u.m. y el contenedor del mayorista B de 300 u.m., calcular cuántos contenedores habrá de comprar a cada mayorista, con objeto de reducir al mínimo el costo de lo solicitado.

**Solución:**

Un frutero necesita comprar contenedores de fruta a dos mayoristas diferentes para satisfacer su demanda, **minimizando el costo total**.

**Ecuación principal:**

$$
\min Z = 150x_{1}+300x_{2}
$$

**Restricciones:**

- **Naranjas:**
  $$
  8x_{1}+2x_{2}\geq 16\;\mathrm{kg}
  $$

- **Plátanos:**
  $$
  x_{1}+x_{2}\geq 5\;\mathrm{kg}
  $$

- **Manzanas:**
  $$
  2x_{1}+7x_{2}\geq 20\;\mathrm{kg}
  $$

In [ ]:
import numpy as np
import scipy.optimize as opt

c=np.array([150,300])
A=np.array([[-8,-2],
            [-1,-1],
            [-2,-7]])
b=np.array([-16,-5,-20,])
opt.linprog(c,A,b)

        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: 1050.0
              x: [ 3.000e+00  2.000e+00]
            nit: 3
          lower:  residual: [ 3.000e+00  2.000e+00]
                 marginals: [ 0.000e+00  0.000e+00]
          upper:  residual: [       inf        inf]
                 marginals: [ 0.000e+00  0.000e+00]
          eqlin:  residual: []
                 marginals: []
        ineqlin:  residual: [ 1.200e+01  0.000e+00  0.000e+00]
                 marginals: [-0.000e+00 -9.000e+01 -3.000e+01]
 mip_node_count: 0
 mip_dual_bound: 0.0
        mip_gap: 0.0

**Resultado:**

El costo mínimo es de $1050\;\mathrm{u.m.}$, comprando al mayorista **A** $3\;\mathrm{contenedores}$ y comprando al mayorista **B** $2\;\mathrm{contenedores}$

---